### ISSUE TITLE
How to exclude a lambdas from Code Coverage?

### ISSUE THREAD

**Issue by AliRafiee1** [dotnet/core#4361]
I'm using the Visual Studio Code Coverage tools on some Unittests against a couple of DotNet Core projects (Asp.net core and Azure Function). 

I have a couple of methods excluded from code coverage via the ExcludeFromCodeCoverage attribute. These methods have lambda expressions which code coverage still wants to include. 

Is there a way to to exclude the lambda inside these methods from code coverage?

For example: it says that x == "test" is not covered.
```
[ExcludeFromCodeCoverage]
public IActionResult About()
{
    var list = new List<string>();
    list.Where(x => x == "test").FirstOrDefault();

    return View();
}

```

**Comment by scalablecory** [comment]
This is likely happening because the lambda body, from an IL perspective, is not part of the `About()` method.

Getting this to work as-is will likely need changes to the code coverage tool and/or IL generation.

@terrajobst do you know who owns this tool?

**Comment by ericstj** [comment]
I bet the code coverage tools are just looking at the metadata.  The compiler generates a nested class in this case which [won't have that attribute on it](https://sharplab.io/#v2:C4LglgNgNAJiDUAfAAgJgIwFgBQyAMABMugHQAyYAdgI4DcOOaAzAQN44GcEdcDaxJACJgAhgHNKAewDOwMAGNpJAMKSYAUwCClERACe0sEoCiAD3kQArhoBiAJ0kBbVRtUA3dXfHqAuj07ILMQAbEQALASaAEaSlsAAFACU/mwpXARuInYEEEbABAC8BJTqAO5E6GHkeQA8xHgAfEn02OnpubIkAOoAFp7q8aaFDQRDBUUARMDqshOJJDZgdrIA8naC6gBmIpYQCYktaVzIAOwt6QC+OBdAA===).

I bet you could add a config file that excluded this by name: https://docs.microsoft.com/en-us/visualstudio/test/customizing-code-coverage-analysis?view=vs-2019#include-or-exclude-assemblies-and-members

That said, there is probably a feature here for code-coverage tools so we should sync up with the owners.  I'm not sure who that would be.

**Comment by ViktorHofer** [comment]
The most popular cross-plat code coverage tool is [coverlet](https://github.com/coverlet-coverage/coverlet) which we also use in dotnet/runtime. This issue should probably be filed against them. cc @tonerdo @MarcoRossignoli

**Comment by MarcoRossignoli** [comment]
@AliRafiee1 are you using Coverlet?that scenario should be covered today, feel free to open an issue there if you need!

**Comment by nohwnd** [comment]
@jakubch1 could you have a look, or delegate to someone who knows CC in our team please?

**Comment by fhnaseer** [comment]
@AliRafiee1 You can add CompilerGeneratedAttribute in the [runsettings ](https://docs.microsoft.com/en-us/visualstudio/test/customizing-code-coverage-analysis?view=vs-2019) to disable code coverage for lambda functions. 
`<Attribute>^System\.Runtime\.CompilerServices\.CompilerGeneratedAttribute$</Attribute>`

Important Note: If you exclude the CompilerGeneratedAttribute attribute, code that uses language features such as async, await, yield return, and auto-implemented properties is excluded from code coverage analysis. To exclude truly generated code, only exclude the GeneratedCodeAttribute attribute.

**Comment by AliRafiee** [comment]
@fhnaseer is that an all or nothing setting? I don't want all the lambda's to be excluded, just some selected ones.

**Comment by fhnaseer** [comment]
@AliRafiee yes, it's an all or nothing setting. If you want to exclude only specific lambda's, then you need to identify those using compiler generated code and add those inside `<Functions>` tag in the runsettings file.
